flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/OperatorStateInputFormat.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.flink.runtime.checkpoint.StateObjectCollection;
+import org.apache.flink.runtime.jobgraph.OperatorInstanceID;
+import org.apache.flink.runtime.state.BackendBuildingException;
+import org.apache.flink.runtime.state.DefaultOperatorStateBackendBuilder;
+import org.apache.flink.runtime.state.OperatorStateBackend;
+import org.apache.flink.runtime.state.OperatorStateHandle;
+import org.apache.flink.state.api.input.splits.OperatorStateInputSplit;
+import org.apache.flink.streaming.api.operators.BackendRestorerProcedure;
+import org.apache.flink.util.CollectionUtil;
+import org.apache.flink.util.IOUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The base input format for reading operator state from a {@link
+ * org.apache.flink.runtime.checkpoint.savepoint.Savepoint}.
+ *
+ * @param <OT> The type of the input.
+ */
+abstract class OperatorStateInputFormat<OT> extends SavepointInputFormat<OT, OperatorStateInputSplit> {
+
+	private final boolean isUnionType;
+
+	private transient OperatorStateBackend restoredBackend;
+
+	private transient CloseableRegistry registry;
+
+	private transient Iterator<OT> elements;
+
+	OperatorStateInputFormat(String savepointPath, String uid, boolean isUnionType) {
+		super(savepointPath, uid);
+
+		this.isUnionType = isUnionType;
+	}
+
+	protected abstract Iterable<OT> getElements(OperatorStateBackend restoredBackend) throws Exception;
+
+	public OperatorStateInputSplit[] createInputSplits(int minNumSplits) throws IOException {
+		OperatorStateInputSplit[] splits = getOperatorStateInputSplits(minNumSplits);
+
+		if (isUnionType) {
+			return subPartitionSingleSplit(minNumSplits, splits);
+		} else {
+			return splits;
+		}
+	}
+
+	private OperatorStateInputSplit[] subPartitionSingleSplit(int minNumSplits, OperatorStateInputSplit[] splits) {
+		if (splits.length == 0) {
+			return splits;
+		}
+
+		// We only want to output a single instance of the union state so we only need
+		// to transform a single input split. An arbitrary split is chosen and
+		// sub-partitioned for better data distribution across the cluster.
+		return CollectionUtil.mapWithIndex(
+			CollectionUtil.partition(splits[0].getPrioritizedManagedOperatorState().get(0).asList(), minNumSplits),
+			(state, index) ->  new OperatorStateInputSplit(new StateObjectCollection<>(new ArrayList<>(state)), index)
+		).toArray(OperatorStateInputSplit[]::new);
+	}
+
+	private OperatorStateInputSplit[] getOperatorStateInputSplits(int minNumSplits) throws IOException {
+		final OperatorState operatorState = getOperatorState();
+
+		final Map<OperatorInstanceID, List<OperatorStateHandle>> newManagedOperatorStates = new HashMap<>();
+		final Map<OperatorInstanceID, List<OperatorStateHandle>> newRawOperatorStates = new HashMap<>();
```

Comment:
nit: maybe add an "unused" to the variable naming, to indicate that this isn't used at all.